### DISCUSSION TITLE
npm bin build error

### DISCUSSION THREAD

**Discussion by muratbilg** [vercel/community#2433] [Bug Report]
### Summary

Hi,

I keep getting this build error: 
`BUILD_UTILS_GET_NODE_BIN_PATH: Running `npm bin` failed to return a valid bin path (code=1, stdout=Unknown command: "bin"To see a list of supported npm commands, run:  npm help, stderr=)`

Does anyone know where to look for a solution?

Thanks in advance.

UPDATE: Having an old version of Vercel CLI on vercel build environment was causing the issue on our side. So, using new Vercel CLI 29.3.3 version fixed the issue.

**Comment by amyegan** [comment]
Hi @muratbilg! I think we need some more context. Can you tell us more about the code that was showing the error? If you have a minimal reproduction or public repo, that would be very helpful for debugging 🙏

**Reply by muratbilg** [reply]
Hi @amyegan , Thanks for your quick response! I found the reason. Overriding builds cause this. When I close the overrides on build it builds it successfully. I moved our build command into package.json for now but I believe it requires further investigation on vercel side.

**Reply by johkade** [reply]
@amyegan
Having the same problem which only occured 6 days, not caused by changes I made to any sort of config or similar.

I'm also unable to re-deploy builds that have succeeded before that - they now fail with the same error. 
Which hints to me that it's unrelated to my code.

I did try upgrading the vercel-cli, still same error though. 

@muratbilg Can you explain further, what you did to fix the issue?

Here's the error I'm getting in the logs

```sh
NOTE: installing packages succeeded 

Detected Next.js version: 12.2.2
--
Error! Running `npm bin` failed to return a valid bin path (code=1, stdout=Unknown command: "bin"
--
BUILD_UTILS_GET_NODE_BIN_PATH: Running `npm bin` failed to return a valid bin path (code=1, stdout=Unknown command: "bin"To see a list of supported npm commands, run:  npm help, stderr=)

```

Thanks in advance for any help 🙌

**Comment by johkade** [comment]
Hi, I wanted to add some more context to my reply before:

Since around 7 days, we're getting the following build error on vercel:

This is the summary shown on vercel:
```sh
Build Failed
Running `npm bin` failed to return a valid bin path (code=1, stdout=Unknown command: "bin" To see a list of supported npm commands, run: npm help , stderr=)
```

This is the actual console output:
```sh
-------> NOTE: packages installed correctly
Done in 315.11s.
--
17:50:38.278 | Detected Next.js version: 12.2.2
17:50:39.043 | Error! Running `npm bin` failed to return a valid bin path (code=1, stdout=Unknown command: "bin"
17:50:39.043 |  
17:50:39.043 | To see a list of supported npm commands, run:
17:50:39.044 | npm help
17:50:39.044 | , stderr=)
17:50:39.566 | Deployment completed
17:50:39.508 | BUILD_UTILS_GET_NODE_BIN_PATH: Running `npm bin` failed to return a valid bin path (code=1, stdout=Unknown command: "bin"To see a list of supported npm commands, run:  npm help, stderr=)
```
Environment:
- node 14 (tried with 16 - same issue)
- next@12.2.2
- yarn workspaces monorepo with two separate next apps, and a react native app

There are a few more weird things about this:
- Builds that succeeded ~2 weeks ago now also fail with this error when re-run, even after clearing all cache on vercel.
- We didn't change any config lately or any code that could have possibly caused it, as far as we're aware of
- the now failing builds, that earlier worked (from same commit-hash), seem to prove that the error is not related to our code
- builds for another project in the workspace with practically the same setup work fine

I've tried everything I could including upgrading vercel, node js, setting the $PATH and more, but can't seem to get it to work.
Also tried renaming build scripts, which didn't help either.

Unfortunately renaming build scripts, as suggested before, did not work in our case.

Would be very thankful for any help!

Thanks in advance 🙌

**Reply by muratbilg** [reply]
Hi @johkade, If you use override feature on build command(find it under settings --> general --> Build & Development Settings), disable it and move your build command into your project scripts. We were using npm run test && npm run build as an override build command but after trying everything I realised that override feature is not working on vercel. So, I moved this command into our package.json file as a temp solution and still waiting reply from @amyegan. I hope it helps 🙌🏻.

**Reply by johkade** [reply]
@muratbilg You're a life safer 🥳 

I could in fact make the build run again by removing the build override command on vercel and moving the custom stuff to a postinstall script in package.json, as you suggested.

So thanks a ton for the hint 🚀 

In conclusion, it seems that build-command-overrides seem to be broken at least under certain circumstances.
So the temporary solution is to disable them on vercel and do any custom logic and custom setup directly in your project via a postinstall script or similar.

**Reply by AgnesGresset** [reply]
Thanks! Also had the same issue and did remove the override feature in the Vercel settings and it did at least fix the deployment. But will be grateful for a fix of the override commands by Vercel!

**Reply by amyegan** [reply]
Thank you all for working together on this solution. I have not been able to reproduce the problem with my test app. Does anyone here have an example repo they can share, or specific steps I can follow to build my own example? I want to get the engineering team everything they need for investigating and fixing any related bugs in our deployment system 🙏

**Reply by johkade** [reply]
Unfortunately I can't share the repo, however here's what we had as the custom build command:
`./scripts/vercel-config.sh && next build` - so just running a simple script, nothing more.

The build didn't even get to the point where the script was actually executed, though.

The content of the script was essentially just this:

```sh
if [ "$VERCEL_ENV" = "production" ]; then
    # init with prod config
elif [ "$VERCEL_ENV" = "preview" ]; then
    # init with staging config
else
    # init with dev config
fi
```

**Reply by muratbilg** [reply]
Hi @amyegan, I don't know what causes it only thing we know overriding triggers it but I believe they can easily detect where it comes from since this line of error shows specific error-code defined in Vercel environment. `17:50:39.508 | BUILD_UTILS_GET_NODE_BIN_PATH: Running `npm bin` failed to return a valid bin path (code=1, stdout=Unknown command: "bin"To see a list of supported npm commands, run:  npm help, stderr=)`

**Reply by AgnesGresset** [reply]
We also have a very simple override file with only one check:

`if [[ $VERCEL_GIT_COMMIT_REF == "develop" ]]; then
  npm run storybook:build:vercel
fi`

Thanks!
